### PR TITLE
feat: add context parallel.

### LINKED ISSUE #43: Context Parallelism (w/ FSDP2 for DTensor)
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

Provide a code snippet on how new APIs/changes would be used by others.

**Describe alternatives you've considered**

**Additional context**

Add any other context or screenshots about the feature request here

### PR DESCRIPTION
**Pull request by joyang-nv**
# What does this PR do ?

This PR adds context parallel with FSDP2 inside nemo rl.
It makes use of torch native experimental context parallel.
This PR has some known issues and will be fixed in following up PR.
1. Use CUDNN_BACKEND when possible for perf optimization.
2. The way to get logits in context parallel enabled case is not optimal.

## Convergence
Run on https://github.com/NVIDIA/NeMo-RL/commit/d77da61899ff8637baf052d31afdfd01de6ade06

Reward improves well.
image

Has little difference on the below three.
image

## Time Cost
image

# Issues
Closes #43

# Usage
Simply set `policy.dtensor_cfg.context_parallel_size` either in command line or yaml file to the desired value to enable it.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/NeMo-RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/NeMo-RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/NeMo-RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 1]
```diff
@@ -262,6 +277,68 @@ def __init__(
...
+
+        return context_parallel(
+            cp_mesh,
+            buffers=cp_buffers,
+            buffer_seq_dims=cp_seq_dims,
+            no_restore_buffers=cp_no_restore_buffers,
+        )
+
+    # Refer to nemo impl. Below is original comment.
+    # based on https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/utils.py#L138
```
Comment (terrykong) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: i know this function may be copied, but do you mind using permalinks to help preserve the link?

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 2]
```diff
@@ -262,6 +277,68 @@ def __init__(
...
+    @staticmethod
+    def get_train_context():
+        """Create a train context.
+
+        Args:
+            enable_loss_parallel (bool): Whether to enable loss parallelism.
+            enable_compiled_autograd (bool): Whether to enable compiled autograd.
```
Comment (terrykong) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: are these args needed?

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 3]
```diff
@@ -262,6 +277,68 @@ def __init__(
...
+        """
+
+        @contextlib.contextmanager
+        def context(cp_context: Optional[Generator[None, None, None]] = None):
```
Comment (terrykong) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: any reason this can't be a staticmethod and contextmanager, in other words fold this into get_training_context so it's something like:

```
@contextmanager
@staticmethod
def attention_context(cp_context: Optional[Generator[None, None, None]] = None)
```

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 4]
```diff
@@ -404,45 +481,107 @@ def train(
...
+                    context_parallel_ctx = (
+                        self.create_context_parallel_ctx(
+                            cp_mesh=self.cp_mesh,
+                            cp_buffers=cp_buffers,
+                            cp_seq_dims=[sequence_dim] * len(cp_buffers),
+                            cp_no_restore_buffers=set(cp_buffers),
                         )
+                        if self.cp_size > 1
+                        else None
+                    )
```
Comment (terrykong) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: this is a little bit of a nit, but could you wrap this section in `if self.cp_size > 1` for readability so it's easier to see that things like `seq_index` are only needed for cp?

File: tests/unit/models/policy/test_dtensor_worker.py [hunk 5]
```diff
@@ -358,7 +364,7 @@ def logprob_setup(request, two_gpu_virtual_cluster):
 
     try:
         config = create_test_config(
-            model_name, tp, cpu_offload, sequence_parallel, activation_checkpointing
+            model_name, tp, 1, cpu_offload, sequence_parallel, activation_checkpointing
```
Comment (yuki-97) [inline tests/unit/models/policy/test_dtensor_worker.py]: can you add something like `test_dtensor_worker_logprob_cp2_matches_no_cp` like `test_dtensor_worker_logprob_tp2_matches_no_tp`?

File: tests/unit/models/policy/test_dtensor_worker.py [hunk 6]
```diff
@@ -291,24 +293,28 @@ def training_setup(request, two_gpu_virtual_cluster):
...
+        (TEST_ASSETS.TINY_QWEN3_MODEL_PATH, 1, 1, True, False, True),
+        (TEST_ASSETS.TINY_QWEN3_MODEL_PATH, 1, 1, False, True, True),
+        (TEST_ASSETS.TINY_QWEN3_MODEL_PATH, 1, 1, True, True, True),
+        (TEST_ASSETS.TINY_QWEN3_MODEL_PATH, 1, 2, False, False, False),
+        (TEST_ASSETS.TINY_GEMMA3_MODEL_PATH, 1, 1, True, True, False),
+        (TEST_ASSETS.TINY_GEMMA3_MODEL_PATH, 1, 1, True, False, True),
+        (TEST_ASSETS.TINY_GEMMA3_MODEL_PATH, 1, 1, False, True, True),
+        (TEST_ASSETS.TINY_GEMMA3_MODEL_PATH, 1, 1, True, True, True),
+        # CP doesn't support gemma3 due to spda input has attent_mask != None.
     ],
```
Comment (yuki-97) [inline tests/unit/models/policy/test_dtensor_worker.py:318]: I feel it's better to add an assert somewhere to remind user when they use gemma3 model with cp>1 in training.

Comment (joyang-nv) [reply tests/unit/models/policy/test_dtensor_worker.py:318]: No, I don't think I can do this. First, I didn't test all gemma3 models, it is not due to specific model arch. Second, I not sure some argument combination could work. Unfortunately, I can't predicate exactly what a model with forward arguments could be supported or not till SPDA is run.

Comment (yuki-97) [reply tests/unit/models/policy/test_dtensor_worker.py:318]: you can check if these two can work: `Gemma3ForCausalLM`, `Gemma3ForConditionalGeneration`.
e.g. `google/gemma-3-1b-it` is `Gemma3ForCausalLM` and can't work.

otherwise user can't know what happen from the error log when they use gemma3 model with cp>1.

Comment (joyang-nv) [reply tests/unit/models/policy/test_dtensor_worker.py:318]: I think the case is much more complex than this single model. But for this single model we could assert according as you suggested.

**Review by terrykong — APPROVED** [review]

**Review by terrykong — DISMISSED** [review]
